### ISSUE TITLE
Consider adding user-defined functions

### ISSUE THREAD

**Issue by katsaii** [katsaii/catspeak-lang#4]
This could potentially cause breaking changes to the compiler and runtime, especially if it becomes possible to call catspeak functions from within GameMaker.

The syntax could resemble the following:
```
set add : fun a b {
    return : a + b
}
```

The two largest issues with adding functions are:
 - Whether Catspeak functions should be callable by GameMaker code
 - How to maintain multiple VM processes if this is the case

If the values returned from Catspeak functions are needed immediately by GML code, then that defeats the purpose of Catspeak being executed over multiple steps.

**Comment by katsaii** [comment]
Having named arguments seems a bit ambitious, so a new syntax might be:
```
add = fun {
  return : arg.[0] + arg.[1]
}
```
This would be similar to how GameMaker did it in the past. This would also open up the doors to variadic argument Catspeak functions and the ability to pass arguments to the top-level script.

An example function to calculate a mean in Catspeak might be:
```
mean = fun {
  n = 0
  for arg.[_] = m {
    n = n + m
  }
  return : n / get_length arg
}
```

**Comment by katsaii** [comment]
Implemented in e6615d0efbe86b8904c6f7c4eff308c36f430c68
